File: Day 1/10/longest_substring.c
#include <stdio.h>
#include <string.h>

int lengthOfLongestSubstring(char *s) {
    int n = strlen(s);
    int maxLength = 0;
    int map[256] = {0};
    int left = 0;
    
    for (int right = 0; right < n; right++) {
        if (map[s[right]] > 0) {
            left = (map[s[right]] > left) ? map[s[right]] : left;
        }
        map[s[right]] = right + 1;
        maxLength = (right - left + 1 > maxLength) ? right - left + 1 : maxLength;
    }
    
    return maxLength;
}

int main() {
    char s[100001];
    printf("Enter a string: ");
    scanf("%s", s);
    
    int result = lengthOfLongestSubstring(s);
    printf("Length of longest substring without repeating characters: %d\n", result);
    
    return 0;
}

File: Day 1/10/longestSubstring.js
const longestSubstring = (s) => {
    let set = new Set();
    let left = 0;
    let maxLength = 0;
    
    for (let right = 0; right < s.length; right++) {
        while (set.has(s[right])) {
            set.delete(s[left]);
            left++;
        }
        set.add(s[right]);
        maxLength = Math.max(maxLength, right - left + 1);
    }
    
    return maxLength;
}

const longestSubstringAlgorithmically = (s) => {
    let map = {};
    let left = 0;
    let maxLength = 0;
    
    for (let right = 0; right < s.length; right++) {
        if (map[s[right]] >= left) {
            left = map[s[right]] + 1;
        }
        map[s[right]] = right;
        maxLength = Math.max(maxLength, right - left + 1);
    }
    
    return maxLength;
}